codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteAssigner.java

```diff
@@ -0,0 +1,437 @@
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unassignedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unassign all bytes.
+            resetAssigned();
+
+            // Clear the streamable bytes.
+            update(new StreamState() {
+                @Override
+                public Http2Stream stream() {
+                    return stream;
+                }
+
+                @Override
+                public int streamableBytes() {
+                    return 0;
+                }
+
+                @Override
+                public boolean hasFrame() {
+                    return false;
+                }
+            });
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unassignedStreamableBytes() {
+            return streamableBytes - assigned;
+        }
+
+        int unassignedStreamableBytesForTree() {
+            return unassignedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class StreamVisitor implements Http2StreamVisitor {
+        boolean visitorDone;
+        Visitor assignmentVisitor;
+        RuntimeException error;
+        boolean inForEachStream;
+
+        void forEachStream(Visitor assignmentVisitor) {
+            assert !inForEachStream;
+            inForEachStream = true;
+            try {
+                this.assignmentVisitor = assignmentVisitor;
+                visitorDone = false;
+                error = null;
+                try {
+                    connection.forEachActiveStream(this);
+                } catch (Http2Exception e) {
+                    // Should never happen since the visitor doesn't throw.
+                    throw new IllegalStateException(e);
+                }
+
+                // If an error was caught when calling back the visitor, throw it now.
+                if (error != null) {
+                    throw error;
+                }
+            } finally {
+                inForEachStream = false;
+            }
+        }
+
+        @Override
+        public boolean visit(Http2Stream stream) {
+            PriorityState state = state(stream);
+            try {
+                if (!visitorDone) {
+                    // Once the user-supplied visitor indicates that it's done, don't call it again.
+                    visitorDone = !assignmentVisitor.visit(stream, state.assigned);
+                }
+            } catch (RuntimeException e) {
+                // Stop calling the visitor, but continue in the loop to reset the assignment for
+                // all remaining states.
+                visitorDone = true;
+                error = e;
+            } finally {
+                state.resetAssigned();
```

Comment:
It seems like the current division is still assuming a 2 phase approach. Allocation first and then writing. Why don't we let the `StreamByteAssigner` completely determine how it allocates and when it writes?  The `StreamByteAssigner.assign` and `StreamByteAssigner.processAssignedBytes` could be combined into a single method. The `StreamByteAssigner` could accept some sort of writer that is used to do the writing.

```java
interface StreamByteAssigner {
  void doAssignment(int maxBytes, SteamByteWriter write);
}

interface SteamByteWriter {
  void writeBytes(Http2Stream stream, int maxBytes);
}
```

I think the biggest benefit of the existing structure is the`StreamByteAssigner` has more control over how it iterates over streams, and isn't necessarily limited to iterating in active stream order. However I think the proposed approach allows at least the same amount of control, may alleviate some of the extra state/boundary conditions with the current iteration approach, and doesn't necessarily assume that things must happen in a two phase process.

> If we don't use a visitor, the flow controller has no choice but to loop over every stream in the connection.

It seems the real limitation is the iteration order. We could still expose a method to terminate the iteration.